velox/functions/prestosql/aggregates/AverageAggregate.cpp

```diff
@@ -223,50 +236,72 @@ class AverageAggregate : public exec::Aggregate {
-          updateNonNullValue(groups[i], count, sum);
+          if constexpr (checkNullFields) {
+            VELOX_USER_CHECK(
+                !baseSumVector->isNullAt(i) && !baseCountVector->isNullAt(i));
+          }
+          updateNonNullValue(
+              groups[i],
+              baseCountVector->valueAt(i),
+              baseSumVector->valueAt(i));
         });
       }
-    } else if (decodedPartial_.mayHaveNulls()) {
-      rows.applyToSelected([&](vector_size_t i) {
-        if (decodedPartial_.isNullAt(i)) {
-          return;
-        }
-        auto decodedIndex = decodedPartial_.index(i);
-        if constexpr (checkNullFields) {
-          VELOX_USER_CHECK(
-              !baseSumVector->isNullAt(decodedIndex) &&
-              !baseCountVector->isNullAt(decodedIndex));
-        }
-        updateNonNullValue(
-            groups[i],
-            baseCountVector->valueAt(decodedIndex),
-            baseSumVector->valueAt(decodedIndex));
-      });
     } else {
-      rows.applyToSelected([&](vector_size_t i) {
-        auto decodedIndex = decodedPartial_.index(i);
-        if constexpr (checkNullFields) {
-          VELOX_USER_CHECK(
-              !baseSumVector->isNullAt(decodedIndex) &&
-              !baseCountVector->isNullAt(decodedIndex));
+      SelectivityVector baseRows = translateToInnerRows(decodedPartial_, rows);
+
+      DecodedVector baseSumDecoded(*baseRowVector->childAt(0), baseRows);
+      DecodedVector baseCountDecoded(*baseRowVector->childAt(1), baseRows);
+
+      if (decodedPartial_.isConstantMapping()) {
+        if (!decodedPartial_.isNullAt(0)) {
+          auto decodedIndex = decodedPartial_.index(0);
+          if constexpr (checkNullFields) {
+            VELOX_USER_CHECK(
+                !baseSumDecoded.isNullAt(decodedIndex) &&
+                !baseCountDecoded.isNullAt(decodedIndex));
+          }
+          auto count = baseCountDecoded.template valueAt<int64_t>(decodedIndex);
+          auto sum =
+              baseSumDecoded.template valueAt<TAccumulator>(decodedIndex);
+          rows.applyToSelected([&](vector_size_t i) {
+            updateNonNullValue(groups[i], count, sum);
+          });
         }
-        updateNonNullValue(
-            groups[i],
-            baseCountVector->valueAt(decodedIndex),
-            baseSumVector->valueAt(decodedIndex));
-      });
+      } else {
+        rows.applyToSelected([&](vector_size_t i) {
+          if (decodedPartial_.isNullAt(i)) {
+            return;
+          }
+          auto decodedIndex = decodedPartial_.index(i);
+          if constexpr (checkNullFields) {
+            VELOX_USER_CHECK(
+                !baseSumDecoded.isNullAt(decodedIndex) &&
+                !baseCountDecoded.isNullAt(decodedIndex));
+          }
+          updateNonNullValue(
+              groups[i],
+              baseCountDecoded.template valueAt<int64_t>(decodedIndex),
+              baseSumDecoded.template valueAt<TAccumulator>(decodedIndex));
```

Comment:
I wonder if its possible to create a utility function for this code since its being repeated quite often with minor differences. Those differences can be enclosed within constexpr checks and templated. What do you think?